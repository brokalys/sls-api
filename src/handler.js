import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
