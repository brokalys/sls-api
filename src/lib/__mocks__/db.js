export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
