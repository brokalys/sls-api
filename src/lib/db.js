import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
