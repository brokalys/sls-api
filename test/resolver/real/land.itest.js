import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import { authenticateAs } from 'helpers';
import db from 'db-config';

const { query } = createTestClient(server);

describe('Resolver: land - real queries from customer apps', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  afterEach(jest.clearAllMocks);

  describe('map app', () => {
    test('retrieve all the historical data', async () => {
      authenticateAs('mapApp', server);

      const response = await query({
        query: `
          query($id: Int!, $filter: PropertyFilter) {
            land(id: $id) {
              id
              bounds
              properties(filter: $filter) {
                results {
                  category
                  type
                  rent_type
                  price
                  calc_price_per_sqm
                  rooms
                  area
                  floor_min: floor
                  date: published_at
                }
              }
              vzd {
                land {
                  date: sale_date
                  price
                  area: land_total_area_m2
                }
              }
            }
          }
        `,
        variables: {
          id: 1,
          filter: {
            price: { gte: 1 },
          },
        },
      });

      expect(response).toMatchSnapshot();
    });

    test('retrieve the historical data from land that has no sale_date', async () => {
      authenticateAs('mapApp', server);

      const response = await query({
        query: `
          query($id: Int!) {
            land(id: $id) {
              id
              bounds
              vzd {
                land {
                  date: sale_date
                  price
                  area: land_total_area_m2
                }
              }
            }
          }
        `,
        variables: {
          id: 2,
        },
      });

      expect(response).toMatchSnapshot();
    });
  });
});
