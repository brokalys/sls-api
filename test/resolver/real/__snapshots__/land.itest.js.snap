// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver: land - real queries from customer apps map app retrieve all the historical data 1`] = `
Object {
  "data": Object {
    "land": Object {
      "bounds": "56.992294 24.136619, 56.976394 23.99579, 56.992294 24.136619",
      "id": 1,
      "properties": Object {
        "results": Array [
          Object {
            "area": 90,
            "calc_price_per_sqm": 1.11,
            "category": "apartment",
            "date": "2021-04-12T19:47:30.000Z",
            "floor_min": 2,
            "price": 100,
            "rent_type": "monthly",
            "rooms": null,
            "type": "rent",
          },
          Object {
            "area": null,
            "calc_price_per_sqm": null,
            "category": "apartment",
            "date": "2021-04-12T19:47:30.000Z",
            "floor_min": null,
            "price": 200,
            "rent_type": "monthly",
            "rooms": null,
            "type": "rent",
          },
        ],
      },
      "vzd": Object {
        "land": Array [
          Object {
            "area": 479000,
            "date": "2020-09-16T00:00:00.000Z",
            "price": 39000,
          },
        ],
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
